protostar/cheatable_starknet/controllers/expect_events_controller.py

```diff
@@ -0,0 +1,166 @@
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Optional
+
+from protostar.starknet import Address, CairoData, Selector, ReportedException
+
+if TYPE_CHECKING:
+    from protostar.testing import Hook
+    from protostar.cairo_testing import CairoTestExecutionState
+    from protostar.cheatable_starknet.cheatables.cheatable_cached_state import (
+        CheatableCachedState,
+    )
+
+
+@dataclass
+class Event:
+    from_address: Address
+    key: Selector
+    data: Optional[CairoData] = None
+
+
+@dataclass
+class EventMatching:
+    pass
+
+
+@dataclass
+class SkippedEventMatching(EventMatching):
+    emitted_event: Event
+
+
+@dataclass
+class AcceptedEventMatching(EventMatching):
+    emitted_event: Event
+    expected_event: Event
+
+
+@dataclass
+class FailedEventMatching(EventMatching):
+    expected_event: Event
+
+
+@dataclass
+class EventMatchingResult:
+    event_matchings: list[EventMatching]
+
+    @property
+    def failed_event_matchings(self):
+        return [
+            matching
+            for matching in self.event_matchings
+            if isinstance(matching, FailedEventMatching)
+        ]
+
+    @property
+    def should_be_accepted(self) -> bool:
+        return len(self.failed_event_matchings) == 0
+
+
+class ExpectEventsController:
+    def __init__(
+        self,
+        test_finish_hook: "Hook",
+        test_execution_state: "CairoTestExecutionState",
+        cheatable_state: "CheatableCachedState",
+    ) -> None:
+        self._test_execution_state = test_execution_state
+        self._test_finish_hook = test_finish_hook
+        self._cheatable_state = cheatable_state
+        self._test_finish_hook.on(self.compare_expected_and_actual_results)
+
+    def add_expected_events(self, expected_events: list[Event]):
+        self._test_execution_state.expected_events_list.append(expected_events)
+
+    def compare_expected_and_actual_results(self):
+        for expected_events in self._test_execution_state.expected_events_list:
```

Comment:
I have trouble understanding why this list is not flattened, why is that the case?

Reply:
Because a user can call `expect_events` multiple times, with many events. 

```python
expect_events(event_A, event_B) # order matters, A must be before B
expect_events(event_C)
```